### PR TITLE
VQVAE + DCGAN on OASIS dataset - Gabriel Russell 46407764

### PR DESCRIPTION
**Pull request by Gabriel-Russell**
# VQVAE + DCGAN implementation on OASIS Dataset

For this project, I have attempted task 8 which entails creating a generative model for one of the outlined datasets using VQVAE or VQVAE2 which has a SSIM of over 0.6. I have chosen to use a VQVAE model on the OASIS brain dataset, while also using a DCGAN for new image generation. An overview of the files and folders within _PatternAnalysis-2023/recognition/VQVAE_Gabriel_Russell_ is as follows:

**Readme_images:** This is a folder containing images produced from the output of the predict.py script and used for the README.md file.
**README.md:** Summarises the background of the project, training process, results and future work.
**dataset.py:** Contains the code required for building datasets for the models
**modules.py:** Contains the code required for building the models and additional functions for plotting visualisations.
**predict.py:** The script that is run for executing the training process, and producing visualisations of all necessary images.
**train.py:** The main training script for calling functions for the VQVAE and DCGAN training processes.
**train_DCGAN.py:** Contains code for initialising and performing DCGAN training process.
**train_VQVAE.py:** Contains code for initialising and performing VQVAE training, validation and testing processes.

Overall, a reasonable image of an MRI scan of a brain was not able to be reproduced, whereby the average SSIM for the test images was approximately 0.492. The max SSIM that was calculated was 0.54. There is much improvement that could be for this project, particularly for the generation side of things if wanting to produce a clear and reasonable image for this dataset.

Thanks,
Gabriel Russell
46407764

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme:
- Project Overview: Good
- Model: Okay, but should make clear the DCGAN acts as the prior for VQVAE. Need to describe how GAN interfaces with VQVAE to generate images
- Data + Preprocessing: Okay, but need details on preprocessing 
- Usage: Missing
- Training / Loss Curve(s): Good
- Result Demonstration: Good
- References: Good

Functionality:
- Reconstruction: Good
- Generation: Okay, can see it is trying to generate a brain

Code:
- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good

Other Comments:
N/A

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable, missing usage and some other info -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
